src/main/java/fi/aalto/cs/apluscourses/model/Module.java

```diff
@@ -0,0 +1,103 @@
+package fi.aalto.cs.apluscourses.model;
+
+import fi.aalto.cs.apluscourses.utils.Event;
+import fi.aalto.cs.apluscourses.utils.StateMonitor;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Collections;
+import java.util.List;
+import org.jetbrains.annotations.NotNull;
+import org.json.JSONObject;
+
+public class Module {
+
+  public static final int ERROR = StateMonitor.ERROR;
+  public static final int NOT_INSTALLED = StateMonitor.INITIAL;
+  public static final int FETCHING = NOT_INSTALLED + 1;
+  public static final int FETCHED = FETCHING + 1;
+  public static final int LOADING = FETCHED + 1;
+  public static final int LOADED = LOADING + 1;
+  public static final int WAITING_FOR_DEPS = LOADED + 1;
+  public static final int INSTALLED = WAITING_FOR_DEPS + 1;
+
+  public final Event stateChanged = new Event();
+  public final StateMonitor stateMonitor = new StateMonitor(this::onStateChanged);
+
+  @NotNull
+  private final String name;
+  @NotNull
+  private final URL url;
+
+  /**
+   * Constructs a module with the given name and URL.
+   * @param name The name of the module.
+   * @param url The URL from which the module can be downloaded.
+   */
+  public Module(@NotNull String name, @NotNull URL url) {
+    this.name = name;
+    this.url = url;
+  }
+
+  /**
+   * Returns a module constructed from the given JSON object. The object should contain the name of
+   * the module and the URL from which the module can be downloaded. Additional members in the
+   * object don't cause any errors. Example of a valid JSON object:
+   * <pre>
+   * {
+   *   "name": "My Module",
+   *   "url": "https://example.com"
+   * }
+   * </pre>
+   * @param jsonObject The JSON object containing information about a single module.
+   * @return A module constructed from the given JSON object.
+   * @throws MalformedURLException  If the URL of the module is malformed.
+   * @throws org.json.JSONException If the jsonObject doesn't contain "name" and "url" keys with
+   *                                string values.
+   */
+  @NotNull
+  public static Module fromJsonObject(@NotNull JSONObject jsonObject, ModelFactory factory)
+      throws MalformedURLException {
+    String name = jsonObject.getString("name");
+    URL url = new URL(jsonObject.getString("url"));
+    return factory.createModule(name, url);
+  }
+
+  @NotNull
+  public String getName() {
+    return name;
+  }
+
+  @NotNull
+  public URL getUrl() {
+    return url;
+  }
+
+  /**
+   * Returns the names of the modules on which this module is dependent.  This method should not be
+   * called unless the module is in FETCHED state or further.
+   *
+   * @return Names of the dependencies, as a {@link List}.
+   * @throws ModuleLoadException If dependencies could not be read.
+   */
+  @NotNull
+  public List<String> getDependencies() throws ModuleLoadException {
+    return Collections.emptyList();
```

Comment:
this `return Collections.emptyList();` is because it still awaits implementation, right?

Reply:
Yes. Same as https://github.com/Aalto-LeTech/intellij-plugin/pull/57#discussion_r379406878. I'll change it to throw an `UnsupportedOperationException`.

Reply:
Or actually NO! I mean, it won't be implemented in this class. It's something that subclass(es) implement. It should be abstract. Whole this class should be abstract! But making it abstract would cause quite a bit changes to unit tests of this class (instead of creating concrete `Module` objects we should either create subclass objects or mocks).

So yes I see this is a bad solution to have class that is practically abstract but is not "officially" abstract. If you want me to change it in this PR, I can do that. It's an easy and straightforward thing to do, just more changed lines of code to this PR.